winch/codegen/src/masm.rs

```diff
@@ -15,6 +15,28 @@ use wasmtime_environ::PtrSize;
 
 pub(crate) use cranelift_codegen::ir::TrapCode;
 
+/// A macro to acquire a scratch register for the given Wasm value type.
+macro_rules! with_scratch {
+    ($self:expr, $val_type:expr, $callback:expr) => {
+        match $val_type {
+            wasmtime_environ::WasmValType::I32
+            | wasmtime_environ::WasmValType::I64
+            | wasmtime_environ::WasmValType::Ref(wasmtime_environ::WasmRefType {
+                heap_type: wasmtime_environ::WasmHeapType::Func,
+                ..
+            }) => $self.with_scratch::<$crate::masm::IntScratch, _>($callback),
+            wasmtime_environ::WasmValType::F32
+            | wasmtime_environ::WasmValType::F64
+            | wasmtime_environ::WasmValType::V128 => {
+                $self.with_scratch::<$crate::masm::FloatScratch, _>($callback)
+            }
+            _ => unimplemented!(),
+        }
+    };
+}
```

Comment:
WDYT about something like:

```rust
trait MacroAssembler {
    fn with_scratch_ty(&mut self, reg_class: RegClass);
    fn with_scratch<T: ScratchType>(...) { self.with_scratch_ty(T::reg_class(), ...) }
}
```

to avoid macros? That way assemblers would implement `with_scratch_ty` under the hood and `with_scratch` as-is today would become a convenience over that method

Reply:
Yeah, makes sense to me, I'll update.

Reply:
Got rid of the macro, but ended with a slightly different version than what you proposed, mainly because I think it make sense to have the wrapper derive the register class and delegate to `Masm::with_scratch<T,_>`, which is what I was trying to solve for.